cmd/openshift-tests/provider.go

```diff
@@ -95,49 +95,16 @@ func decodeProvider(provider string, dryRun, discover bool) (*exutilcluster.Clus
 		if len(providerInfo.Type) == 0 {
 			return nil, fmt.Errorf("provider must be a JSON object with the 'type' key")
 		}
-		var cloudConfig e2e.CloudConfig
```

Comment:
Doing this would mean we have to be backwards compatible forever in upstream test context (the config for a step is ignorant of the version of openshift it’s testing).  If upstream changes the struct in a non backwards compatible way the rebase will require all the test steps/config in CI to be forked).

I’m not opposed to the change in theory but I’d want some evidence we have a way that is not “openshift-tests api surface to CI changes” before moving forward.

Reply:
No, it's the other way around; _currently_ we expose the upstream API in our API. After this PR we do not.

That is, before this commit we do:
  - unmarshal `provider` into an `e2e.CloudConfig`
  - copy the `e2e.CloudConfig` into an `exutilcluster.ClusterConfig` element-by-element
  - later, copy the `exutilcluster.ClusterConfig` _back_ into an `e2e.CloudConfig`, element-by-element, for the tests to use (https://github.com/openshift/origin/blob/ff70d2d5/cmd/openshift-tests/provider.go#L31)

after this PR we do:
  - unmarshal `provider` into an `exutilcluster.ClusterConfig`
  - later, copy the `exutilcluster.ClusterConfig` into an `e2e.CloudConfig`, element-by-element

Reply:
We're backward compatible on the json format all the way back to 4.1 still, right?

Reply:
Yes. The JSON-parsing unit test is added in the first commit, and none of the input values of those tests is changed after that. (The _output_ values change slightly to reflect two bugfixes: we now correctly set the default vales of NumNodes and MultiMaster even on bare metal; and we no longer ignore the MultiMaster and MultiZone fields in the input JSON if it's trying to override them.)

If you want to add additional examples from openshift/release to the unit tests I can do that